src/rust/bitbox02-rust/src/async_usb.rs

```diff
@@ -77,15 +82,26 @@ where
 /// If this spin finishes the task, the state is moved to
 /// `ResultAvailable`, which contains the result.
 pub fn spin() {
-    let mut state = USB_TASK_STATE.0.borrow_mut();
-    match *state {
-        UsbTaskState::Running(ref mut task) => {
-            let result = spin_task(task);
-            if let Poll::Ready(result) = result {
-                *state = UsbTaskState::ResultAvailable(result);
+    // Pop task before polling, so that USB_TASK_STATE does not stay borrowed during the poll.
+    let mut popped_task = match *USB_TASK_STATE.0.borrow_mut() {
+        // Illegal state, `None` is only valid during the poll.
+        UsbTaskState::Running(None) => panic!("task not found"),
+        // Get the task out, putting `None` in. This allows us to release the mutable borrow on the
+        // state.
+        UsbTaskState::Running(ref mut task @ Some(_)) => core::mem::replace(task, None),
```

Comment:
core::mem::replace seems to be the only way were you can move a non-Copy type without overhead :shrug: the other way would be to use a `Rc<Task>`, so one can have an intermediate step where there are two references before removing one again.

Reply:
I didn't even know that mem::replace was doing something so complicated.. https://stackoverflow.com/questions/27197938/memreplace-in-rust

Reply:
Clippy really is quite good!

https://rust-lang.github.io/rust-clippy/master/index.html#mem_replace_option_with_none

Reply:
yeah, they have some cool lints